integration of code-folding extension into brackets

thehogfather opened the pull request:
Integration of core code-folding extension files into brackets' default extension folder.

thehogfather committed:
integration of code-folding extension into brackets

Florian-R commented:
Out of curiosity, any reason to not keep this as an extension?

prksingh commented:
@Florian-R There has been demand to have code folding in the default user experience for brackets. It made sense to reuse an already popular extension to provide this support in core

abose commented:
@thehogfather Thanks :+1: 
The Travis CL builds are failing mostly due to JSLint related errors. Could you fix the lint errors and update?

Florian-R commented:
@prksinghI had hoped that Brackets has kept some sort of Node philosophy, "small core, vibrant userland" but anyway thanks for your input.

thehogfather commented:
@abose sure thing.

thehogfather committed:
JSHint error fixes

ryanstewart commented:
@Florian-R in general, that's our philosophy, but code folding has been on our list as "core" for a very long time. Most major editors have it out of the box. It's also far and away one of the most popular extensions, so it was clear that users were missing it.

ryanstewart commented:
A couple of things we want to do as part of the merge:
- Double check the shorts that are used. I think they're probably fine but we should take a look at that.
- Move the preferences into our regular preferences system (brackets.json)
- Create an option to disable code folding as preference

Open question here: Do we allow separate code folding preferences per project? or just globally? or per file?

MiguelCastillo commented on src/extensions/default/CodeFolding/DefaultSettings.js:
```diff
@@ -0,0 +1,19 @@
+		alwaysUseIndentFold: true,
+		enableRegionFolding: true,
+		fadeFoldButtons: false,
+        maxFoldLevel: 2 // this value is only used when fold all is called
```
NIT: indentation

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,72 @@
+        folds = "folds";
+    
+    function simplify(folds) {
+        if (!folds) { return folds; }
```
Tweak this to 
```
if (!folds) {
    return folds;
}
```

le717 replied:
But how can you return `folds` if it not passed?

thehogfather replied:
That's equivalent to returning undefined. I've updated for clarity :).

MiguelCastillo commented on src/extensions/default/CodeFolding/DefaultSettings.js:
```diff
@@ -0,0 +1,19 @@
+ */
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
+/*global define*/
+define(function (require, exports, module) {
```
Files are generally mixed tabs/spaces. For consistency, we should stick to 4 spaces.

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,72 @@
+    function simplify(folds) {
+        if (!folds) { return folds; }
+        var res = {}, range;
+        Object.keys(folds).map(function (line) {
```
Why call `map` when you are not really transforming data and creating a new array form it?  A forEach would suffice.

thehogfather replied:
Indeed that is a typo. Should be a forEach.

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    function inflate(folds) {
+        if (!folds) { return folds; }
```
Expand this out to multiple lines

```
if (!folds) {
    return folds;
}
```

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    module.exports = {
+        get: function (id) {
```
Common style is to make these named functions that you assign to the `module.exports`.  E.g.

``` javascript
function get(id) {
}

module.exports.get = get;
```

We probably also benefit from some jsdocs :)

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,72 @@
+            stateManager.set(folds, store);
+            stateManager.save();
+        },
+		getSetting: function (key) {
```
Indentation here wont be a problem once they are moved from here to named functions.

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldcode.js:
```diff
@@ -0,0 +1,217 @@
+            if (typeof pos === "number") {
+                pos = CodeMirror.Pos(pos, 0);
+            }
+            var finder = (options && options.rangeFinder) || CodeMirror.fold.auto;
```
instead of checking is options exists in each line that uses `options`, you can have give it a default value.

``` javascript
options = options || {};
```

MiguelCastillo commented:
@thehogfather  It is awesome that you are brining this into brackets!! High-5.

I did not realize how big your extension was!  But I did a quick pass and it seems that one glaring inconsistency is the mix of spaces/tabs and that lots of the function are defined right in the `module.exports`.  I am guessing it will be a couple of rounds of tweaks/refactorings to get your extension in line with the rest of brackets. :)

Really happy this is coming in.

thehogfather commented:
@MiguelCastillo No problem. I'll work on these fixes and update. :)

thehogfather committed:
Refactored extension to use brackets style exports properties. Added some api documentation for functions.

thehogfather committed:
Moved preferences to brackets.json

thehogfather committed:
Added support for disabling the extension from the preferences dialog.

MarcelGerber commented:
You should probably merge your translations (at least the English ones) into our main `strings.js` as other default extensions do the same.

thehogfather committed:
Merge commit 'e56192e09ceed72834985f2483ed70e339e5c519' into codefolding

* commit 'e56192e09ceed72834985f2483ed70e339e5c519':
  Logos are now perfectly centered
  Whitespace trimming suggested by @MarcelGerber
  Fix some moved APIs that were left over from PR #10641
  Text is now perfectly centered

thehogfather committed:
Merged english translation strings into nls/root/strings.js

thehogfather commented:
I've tidied up and refactored a bit based on the suggestions (thanks all). I've moved preferences to brackets.json and merged the english strings into brackets main strings. There is also an option to disable the extension in the settings dialog (should this be on the menu instead - or as well?). 

I have translations for fi, fr, ja, nl, pt-br and ru but haven't merged yet. Shall I go ahead and merge those?

nethip commented:
@thehogfather Please go ahead and merge all the translations for this, barring fr and ja. The fr and ja will need to go through our translation process.

thehogfather committed:
Merged language strings for fi, nl, pt-br, ru

abose commented:
@thehogfather As i understand, the translations, the HTML dialogs and the menu entries are for the settings dialog for code folding. But as @ryanstewart pointed out in an earlier comment, Brackets doesn't set app preferences with dialogs; But customizable options are set in the preerences file and the default options are set in the 'brackets.json' file in the repo. As the code folding extension is directly integrated with brackets, i think it would be best if it confirms to brackets preferences settings. 
So i think we should
> delete all the preferences dialogs and related strings and move all configurable settings to brackets.json.
> Remove the menu entry for code folding settings
> Code folding by default will be enabled for all users. The option to disable code folding can be set as a preferences entry.
@peterflynn @ryanstewart

MarcelGerber commented:
It's probably easier to have a common prefix for all string names, like `CODE_FOLDING_*`.
Also, you can now remove your old nls files.

thehogfather commented:
@abose thanks for the clarification. I'd misunderstood @ryanstewart's reference to brackets.json. Shall update.

abose commented:
@thehogfather I see a console error message "Cannot assign Ctrl-Alt-X to e4b.TOGGLE_PANEL. It is already assigned to codefolding.expand"- when we run brackets+extract builds. Could we change the expand and collapse shortcuts to something like  "ctrl + shift + ["   and "ctrl + shift + ]"  (like in sublime).

thehogfather commented:
@abose we can certainly change the shortcut keys - although probably not to `Ctrl+Shift+[` since that clashes with `navigate.prevDoc`. May I suggest `Ctrl+Alt+[` and `Ctrl+Alt+]`?

thehogfather committed:
Removed htmltemplates related to dialog boxes for code-folding settings.
Updated shortcut to Ctrl+Alt+[ and Ctrl+Alt+] for expand and collapse

thehogfather commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,127 @@
+        @returns {string} the setting with the specified key
+    */
+    function getSetting(key) {
+        return prefs.get(key, PreferencesManager.CURRENT_PROJECT);
```
@abose @ryanstewart I need some clarification on how to retrieve preferences from the correct scope. For the sake of argument, the above line is attempting to retrieve code-folding settings that have been entered at the project level (i.e., using .brackets.json in a project dir). But it appears that the values in that file are actually not being returned in this instance. In fact after a little debugging, the project scope does not appear in the scope order used in the _getContext function defined in PreferencesManager (perhaps because there is no path property in the context sent from PreferencesManager.CURRENT_PROJECT). What is the correct way to use this api?

MiguelCastillo replied:
@thehogfather `prefs.get(key)` is supposed to return the value by first searching the project `.brackets.json`, and if the key isn't found there, it will search the use (global) `brackets.json`.  You don't need the second parameter.  But it seems like you are after something a bit different.  Can explain on what you are trying to achieve?

thehogfather replied:
@MiguelCastillo thanks, that was how I thought it worked. All I am trying to do is to override any code-folding preferences set at the user or default level with one set at the project level. For instance placing `    "code-folding.enabled": false
` in .brackets.json file to disable code-folding. I couldn't get this to work consistently yesterday so I thought I'd 'force' it to look at the project scope using the second parameter [as suggested by the wiki](https://github.com/adobe/brackets/wiki/Preferences-System#looking-up-prefs). At the moment, when I put a breakpoint in developer console to debug, it works as expected. Without the breakpoint it doesn't seem to be reading the correct preference value.
Any insights or questions would be useful. I'll also have a closer look later today.
Cheers.

thehogfather replied:
Actually all is working consistently and as expected now.

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,127 @@
+        });
+        return res;
+    }
+    /**
```
add a courtesy new line right here :D

abose commented:
shortcuts Ctrl+Alt+[ and Ctrl+Alt+] :+1:

thehogfather committed:
updated to fix preferences so that fold gutter is correctly rendered when projects are changed.

abose commented on src/nls/ru/strings.js:
```diff
@@ -549,21 +549,39 @@ define({
+    "DOCS_MORE_LINK"                            : "Подробнее\u2026",
+
+    //extensions/default/CodeFolding
+    "COLLAPSE_ALL"                  : "Свернуть все",
```
As we are not using the preferences dialog, do we need these translations now? We could just use the root English string as it used only for setting the preferences variable.

abose commented on src/nls/fi/strings.js:
```diff
@@ -624,7 +624,25 @@ define({
+    "DOCS_MORE_LINK"                            : "Lue lisää",
+
+    //extensions/default/CodeFolding
+    "COLLAPSE_ALL"                  : "Pienennä kaikki",
```
As we are not using the preferences dialog, do we need these translations now? We could just use the root English string as it used only for setting the preferences variable.

thehogfather replied:
@abose 5 of the translations are for the menu items under `View` i.e., 'collapse all', 'expand all', 'collapse current', 'expand current', 'collapse custom regions'. I think it is good idea to keep these menu items (as they also show the shortcut keys). If there are strong reasons why you think we should remove them then we can discuss them.

As for the other preference strings, you are absolutely right. I've only kept them to use in the options parameter (name and description) when defining preferences. So they can be taken out of the non root strings.

abose replied:
What does collapse custom regions do?
Also putting in all the code folding options in the top level view menu would increase the menu real estate. It would be ideal to move it into a top level 'code folding' entry in the view/edit menu(as in sublime ) and then put the folding options as submenus. The string freeze is by 6'th, so we would need to put in all strings into brackets before that.

thehogfather replied:
@abose collapse custom regions closes all custom defined region blocks in the current editor. These are defined using comments of the current mode with the keywords `region & endregion` e.g.,:

    \\region
    var point = {x: 1, y: 2};
    \\endregion

With respect to menus,  as far as I know, brackets does not implement submenus yet. Hence I'm not sure how to go about creating the sublime-like sub-menu for code-folding. So did you mean a top level "code folding" entry (just like view/edit), whose menu items are the folding options?

And as for the strings, I have no more changes to make. Are you happy with the code-folding strings present in the root string? In addition, do let me know if you would like me to remove all code-folding related translations I have merged into non-root strings. These can certainly be sorted promptly.

abose replied:
I failed to notice that brackets didn't have submenus; i guess this is the way for now.

peterflynn replied:
@abose See https://trello.com/c/EwLGRkYe/541-native-submenus.  The only reason we don't have it is that it's a sizeable chunk of work: we'd have to decide on the JS `Menus` API, implement submenus natively on Mac and Win, _and_ implement them HTML for Linux / in-browser before the feature could land in master.

abose replied:
OK, got it. we'll have to address it some day.

abose commented:
@thehogfather How would it behave if the user had already installed the extension from the registry. I did a basic round of testing and coudnt find any issues in the case. But will this 'double loading' of the same extension source cause any problems? How do we handle this case?

abose commented:
The package.json file which has the details for the extension description,author,Contributors etc.. is missing. Could you put that file file back in? 
Also need to change the name to "name": "code-folding" from "name": "brackets-code-folding" - to differentiate between the default code folding extension and the code folding extension in the registry. Should the version number could be retained to get which version of the extension we integrated?

MarcelGerber commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,128 @@
+        foldsKey                = "folds";
+
+    //default preference values
+    prefs.definePreference("enabled", "boolean", true,
```
I don't think any other Brackets code uses the fourth parameter `options` just yet.

thehogfather replied:
No I didn't think so either. But I thought if the API exists why not use it?

MiguelCastillo replied:
I like the extra meta data in there :)

thehogfather commented:
@abose at the moment what happens is that the first version of the extension loaded takes precedence. However event handlers on the EditorManager, DocumentManager and ProjectManager may be handled twice. The only drawback I can see is  that persistence of folded states is done twice when closing the application or switching projects. In both cases the integrity of the fold states is preserved.

I can update the version on the registry so that it is only initialised if code-folding doesnt already exist (e.g. by inspecting properties on the CodeMirror object). For those who do not update the extension, the menu items under `View` will still show an entry for the `Settings Dialog`.

thehogfather committed:
added package.json
removed deprecated warnings on event handlers

le717 commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+        which is defined in the preferences. Levels refer to fold heirarchies e.g., for the following
+        code fragment, the function is level 1, the if statement is level 2 and the forEach is level 3
+
+            function sample() {
```
If this (`sample()`) is dead code, can you remove it? :)

le717 commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    /**
+        Collapses all custom regions defined in the current editor
```
I think there's a small indentation issue here.

le717 commented:
One thing I'm noticing is lack of [JSDocs](https://github.com/adobe/brackets/wiki/Brackets-JSDocs-Guidelines). I find the FuncDocr and Reasonable Comments extensions help make that easier. :)

abose commented:
@thehogfather Yes it would be good if the extension is aware that there could be code folding extension by default and do the loading changes accordingly. You could push in a later update to the original extension.

abose commented:
@thehogfather Will be merging this by tomorrow :+1: Thanks for the help. Will have to write the unit tests and some other changes, but once its in master we could help with those.

thehogfather commented:
@abose no problem. Happy to help out anytime.

MiguelCastillo commented:
@thehogfather sorry i disappeared.  Let me take another pass at the code later this evening.

My take on the code folding extension in the registry is that once this is merged in there isn't really a need for the one in the registry, especially if it is **just duplicate functionality**.  In general, I am not too concerned if someone decides to install an extension for functionality that already exists natively in brackets.  However, we should be good citizens and make sure that the integrated extension does not use the same settings as the one that already exists in the registry - to avoid collision.  Currently both use `"code-folding"`, so we would probably pick another name for this integrated extension so that people with the extension already installed don't have conflicts.

Also, I am thinking that this will be announced as a feature for a future release, so lots of people will be aware that code folding is native in brackets :D

@abose default extensions don't need a package.json and the couple that have it are theme extensions because the `theme` field is needed by the extension manager.  So package.json file should *not* come in - to stay consistent with the other default extensions.  I don't believe there is a convention where default extensions need to have a name that starts with "brackets" - I don't think it is a requirement either.

abose commented:
@MiguelCastillo The name for the extension in the registry is 'brackets-code-folding' and the one that is integrated is 'code-folding' to prevent the conflicts as you pointed out. Since the extension is used by a large number of users, I was hoping that @thehogfather would update the extension in the registry to  consider the case where brackets already have the extension as default[detected using the extension name from package.json?]. If the extension is pulled out of the registry, wouldn't It affect the users who have the extension installed. That was why I thought to put in the package.json file in; apart from the dev attributions. Also most of the default extension (except themes) were developed inside the brackets source base itself; that may be why the package details are missing in them.

Also did you get a chance to look at the code, I was planning to integrate it today for kicking off the string translations and giving sufficient window for unit tests from our side as the 1.3 release is very close now.

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,114 @@
+define(function (require, exports, module) {
+    "use strict";
+    var PreferencesManager      = brackets.getModule("preferences/PreferencesManager"),
+        prefs                   = PreferencesManager.getExtensionPrefs("code-folding"),
```
Any objection to renaming `code-folding` to maybe `brackets-code-folding`? Simply to helps everyone keep settings separate.

thehogfather replied:
in principle no objection. But I would suggest that I change the prefs in the version in the registry to "brackets-code-folding" for consistency - to match the name in the package.json.

abose replied:
That would be great. It would also help existing users keep saved settings from the extension.

MiguelCastillo replied:
@thehogfather Are you proposing that the integrated and the one in the registry should have the same preferences name?

thehogfather replied:
@MiguelCastillo no. Just proposing that the preference name should match the name in the package.json. So currently the version in the registry stores preferences using the "code-folding" key. I'll update it to "brackets-code-folding" and leave the version integrated into brackets as "code-folding". This also means any preferences saved in the old extension will be kept in the integrated version.

MiguelCastillo replied:
Gotcha.  That makes sense.  I was just curious if we wanted users to just transparently use the settings they have.  In which case I expect them to remove the extension.

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,114 @@
+    prefs.definePreference("folds", "object", {});
+
+    /**
+        Simplifies the fold ranges into an array of pairs of numbers.
```
@thehogfather - do you think we can get all the jsdocs to follow the conventions followed in Brackets?

```
/**
  *  Simplifies the fold ranges into an array of pairs of numbers.
  *
  * @param {!{number: {from: {ch, line}, to: {ch, line}} folds the raw fold ranges indexed by line numbers
  * @returns {number: Array<Array<number>>} an object whose keys are line numbers and the values are array
  *        of two 2-element arrays. First array contains [from.line, from.ch] and the second contains [to.line, to.ch]
  */
```

[This](https://github.com/adobe/brackets/blob/master/src/extensions/default/JavaScriptCodeHints/main.js#L122) is a example.

thehogfather replied:
@MiguelCastillo not a problem. Thanks for the pointer.

MiguelCastillo commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,114 @@
+        @param {string} path the document path
+        @returns {number: {from: {line, ch}, to: {line, ch}}} the line folds for the document at the specified path
+    */
+    function get(path) {
```
Could we rename this `get` and `set` to `getFolds` and `setFolds`?  It improves intent and readability. :)

thehogfather replied:
Sure - I agree.

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldcode.js:
```diff
@@ -0,0 +1,226 @@
+            return;
+        }
+
+        var myWidget = makeWidget();
```
Convention in Brackets is that variables are declared/defined at the beginning of the method scoping them.

peterflynn replied:
@MiguelCastillo Fwiw, not necessarily true. Because of JSLint, it's safe (often actually saf<i>er</i>) to declare vars further down in a function at the point when they're initialized.  We do have plenty of code in core where vars are not all grouped at the top of the scope.

MiguelCastillo replied:
Hmmm, the only reason I say this is because JSLint would complain if your vars werent declared at the top of your functions.  I am guessing this changed.  Good to know though. :)

peterflynn replied:
We use the `vars` JSLint option, which lets you declare vars anywhere in the scope -- as long as they're not duplicated or referenced before the declaration.  It's sort of a best of both worlds, because it effectively makes var declarations work more like Java or C++ (assuming you don't ignore lint errors).

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -0,0 +1,308 @@
+        var fade = prefs.getSetting("fadeFoldButtons");
+        var gutter = $(cm.getGutterElement());
+        var i = from;
+        var isFold = function (m) {
```
Can we expand `isFold` and `clear` function expressions to function declarations?  I havent spotted more, but function expressions aren't really used in Brackets much.

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -0,0 +1,308 @@
+            @param {Number} line the
+            @return {from: {ch: Number, line: Number}, to: {ch: Number, line: Number}} the range that hides the specified line or undefine if the line is not hidden
+        */
+        function _isCurrentlyFolded(line) {
```
Functions that start with `_` are conventionally private...  Can we add `@private` to the jsdoc?

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -0,0 +1,308 @@
+            @return {from: {ch: Number, line: Number}, to: {ch: Number, line: Number}} the range that hides the specified line or undefine if the line is not hidden
+        */
+        function _isCurrentlyFolded(line) {
+            var keys = Object.keys(cm._lineFolds), i = 0, r;
```
Can we please name `r` to a little more descriptive name?  It helps us mere humans read the code a bit better :)

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -0,0 +1,308 @@
+
+    function updateInViewport(cm, from, to) {
+        var vp = cm.getViewport(), state = cm.state.foldGutter;
+        from = !isNaN(from) ? from : vp.from;
```
NIT: double negation tends to be harder to read.

``` javascript
from = isNaN(from) ? vp.from : from;
to = isNan(to) ? vp.to : to;
```

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -0,0 +1,308 @@
+            }
+            state.from = changeObj.from.line;
+            state.to = 0;
+            window.clearTimeout(state.changeUpdate);
```
If we use `lodash` we can clean this setting/clearing timeout with the debounce function.  Very nifty little function. https://lodash.com/docs#debounce

peterflynn replied:
@MiguelCastillo Though I wonder if it'd be better to try to keep this as close as possible to the original CM code, with the goal of getting our few changes merged back upstream so we don't have to maintain a fork of it...

MiguelCastillo commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -0,0 +1,308 @@
+    */
+    function onViewportChange(cm) {
+        var state = cm.state.foldGutter;
+        window.clearTimeout(state.changeUpdate);
```
Same `debounce` suggestion as above

MiguelCastillo commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+        ProjectManager          = brackets.getModule("project/ProjectManager"),
+        KeyBindingManager       = brackets.getModule("command/KeyBindingManager"),
+        ExtensionUtils          = brackets.getModule("utils/ExtensionUtils"),
+        Menus					= brackets.getModule("command/Menus"),
```
NIT: A little indentation detail

MiguelCastillo commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+    //still using slightly modified versions of the foldcode.js and foldgutter.js since we
+    //need to modify the gutter click handler to take care of some collapse and expand features
+    //e.g. collapsing all children when 'alt' key is pressed
+    var foldGutter              = require("foldhelpers/foldgutter"),
```
Move these require statements along with the other one in line 51.

MiguelCastillo commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+        Initialise the extension
+    */
+    function init() {
+        if (!CodeMirror.fold.combine && _prefs.getSetting("enabled")) {
```
In the name of readability, can we make this `if` statement be an early `return`?  This is merely to get rid of the deep indentation in the body of the `if` statement.

MiguelCastillo commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+        KeyBindingManager       = brackets.getModule("command/KeyBindingManager"),
+        ExtensionUtils          = brackets.getModule("utils/ExtensionUtils"),
+        Menus					= brackets.getModule("command/Menus"),
+        _prefs                  = require("Prefs"),
```
Why underscore this `_prefs` but not in an other file/module?

MiguelCastillo commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,326 @@
+
+    function createGutter(editor) {
+        var cm = editor._codeMirror;
+        if (cm) {
```
Can we also do an early `return` here?

MiguelCastillo commented:
@thehogfather this code is definitely landable right now and it is looking very good!

There are a few changes that seem manageable to make the code fit better with brackets style.  I think we are almost there :)

thehogfather committed:
updated JSDocs and general code clean up.

thehogfather committed:
Merge commit 'fd6c8acca823e3cf197e6d3ac0ab9adea190843e' into codefolding

* commit 'fd6c8acca823e3cf197e6d3ac0ab9adea190843e': (31 commits)
  Added a newline
  Adding newline
  Updating the comments
  Updating the comments
  Removing this preference as it should be set to true by default.
  Changing function name.
  Decreasing timeout to 4 sec.
  Adding unit test cases. Submitting changes suggested in PR review.
  Setting ONE_DAY to correct value. Initially value used for testing was left by mistake. Removing serverTestURL as it is not needed now.
  Moving lastTimeSentData before making AJAX call to make sure data is not sent before ONE_DAY as setting it to always() will be asynchronous
  Move the implementation to the extensions. Data will be send to the server as soon as user has launched the Brackets and dialog box prompting for user to opt-in and opt-out is closed. Rework the comments and defects. Test cases has not been added yet. Will be adding to the next PR.
  Updated by ALF automation.
  Updated by ALF automation.
  Extended {APP_NAME}
  Correcting jshint errors.
  Deleting duplicate string.
  Synchronized with latest root commit
  "developer" -> "pengembang"
  Fixed text wrapping
  Add comments..
  ...

Conflicts:
	src/nls/root/strings.js

thehogfather commented:
@MiguelCastillo just pushed an update for all those fixes + merged with the latest commit from master.

abose commented:
@thehogfather  Thanks for the help. Merged into master.
Any suggestion on the unit tests that needs to be added?

MiguelCastillo commented:
@thehogfather yeah man, your last changes look very good. Thank you, this is a great addition to brackets!  Now you gotta come up with another really cool extension :)

@abose I really like the idea of adding the package.json for attribution :)

le717 commented:
![Dancing banana](http://rack.3.mshcdn.com/media/ZgkyMDEyLzEwLzE5LzExXzMzXzMzXzE3Nl9maWxlCnAJdGh1bWIJMTIwMHg5NjAwPg/462b8072)

Congrats on the merge, @thehogfather! :smile:

thehogfather commented:
:+1: thanks all. Glad I could help.

@abose I guess I would mainly check that folded states are persisted and restored correctly, and that the preference options  (enabled, minFoldSize etc.) work as intended.

peterflynn commented on src/extensions/default/CodeFolding/foldhelpers/foldcode.js:
```diff
@@ -0,0 +1,268 @@
+/**
+ * Based on http://codemirror.net/addon/fold/foldcode.js
+ * @author Patrick Oladimeji
```
I think we still need to keep the original copyright from CodeMirror here

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,352 @@
+        "i18n" : "lib/i18n"
+    },
+    locale: brackets.getLocale()
+});
```
This should be unnecessary now that this is a part of core

peterflynn commented on src/extensions/default/CodeFolding/strings.js:
```diff
@@ -0,0 +1,38 @@
+
+    module.exports = require("i18n!nls/strings");
+
+});
```
Ditto -- I think this whole file is no longer used

peterflynn commented on src/extensions/default/CodeFolding/foldhelpers/latex-fold.js:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Fold latex code regions
```
I wonder if a better approach would be to make this more extensible -- an important goal of the LanguageManager work is to make it so any feature in Brackets can be enabled for additional languages just by installing an extension.  Latex and the custom regions feature both seem like potential candidates to move out into extensions.

Custom regions also adds a menu item that I suspect few users will understand, and that's concerning to me.  Since refactoring this to be more extensible is probably a post-1.3 thing, I wonder if it'd be better to just disable the custom regions feature for now?  Or have the menu item hidden by default, and accessible only by setting a preference?

thehogfather replied:
@peterflynn I agree - and I've already tested moving the latex-fold function into the latex extension and that is works as expected. They can be safely removed from the main.js file (if you haven't sealed changes to 1.3 yet).
I also think the custom region should be disabled by default (and the menu option hidden).

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,116 @@
+      * @param {Object} folds the fold ranges to save for the current document
+      */
+    function setFolds(path, folds) {
+        store[path] = simplify(folds);
```
This _really_ should be in view state and not the main preferences file -- it's exactly the kind of thing view state was intended for.  This seems like something we should fix for 1.3 lest we start polluting everyone's preferences file (which should be easily human-readable) with folding state.

peterflynn replied:
Filed as #10846 to make sure we track this

abose replied:
@thehogfather #10846 Instead of using the prefs , could we migrate the folding state to state.json?

thehogfather replied:
@abose yes - for the folds. I suspect the other settings would need to be in the regular prefs (brackets.json).

abose replied:
@thehogfather could use PreferencesManager.getViewState("codefolding.states"); and PreferencesManager.setViewState("codefolding.states", state) to put to the state.json file.

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,116 @@
+      * Clears all the saved line folds for all documents.
+      */
+    function clearAllFolds() {
+        prefs.set(foldsKey, {});
```
This function appears to be unused, but I think it also has a bug: this doesn't reset the `store` var used elsewhere, so any call to `setFolds()` later seems like it would immediately restore _all_ the folds that had been cleared previously.

It's probably safest just to re-get the pref before setting it each time.  There may be other bugs lurking with the current approach too (e.g. if you edit the prefs file... does it blow those changes away?)

thehogfather replied:
@peterflynn yes editing the prefs file for documents that are currently open in brackets (has no effect - the changes are restored one the document becomes the active editor). This is a bug - that  your safe approach should fix.

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,116 @@
+                           {name: strings.ALWAYS_USE_INDENT_FOLD, description: strings.ALWAYS_USE_INDENT_FOLD_HELP});
+    prefs.definePreference("enableRegionFolding", "boolean", true,
+                           {name: strings.ENABLE_REGION_FOLDING, description: strings.ENABLE_REGION_FOLDING});
+    prefs.definePreference("fadeFoldButtons", "boolean", false,
```
I wonder if this preference should be on by default.

I can't actually get it to work though -- when I set it to true, it seems to have no effect (at least on JS or LESS files).  Will file a bug.

peterflynn replied:
Ok, false alarm with the preferences issue.

I wonder if the best UX is actually in between this setting (hidden until mouseover) and the default (always visible) -- having the triangles be fainter until you mouse over the gutter, then getting darker.  I may mock that up and see how it feels...

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -0,0 +1,116 @@
+    prefs.definePreference("saveFoldStates", "boolean", true,
+                           {name: strings.SAVE_FOLD_STATES, description: strings.SAVE_FOLD_STATES_HELP});
+    prefs.definePreference("alwaysUseIndentFold", "boolean", true,
+                           {name: strings.ALWAYS_USE_INDENT_FOLD, description: strings.ALWAYS_USE_INDENT_FOLD_HELP});
```
What does it mean to have this enabled?  What is the behavior e.g. in a JS file if this conflicts with the standard brace-fold/comment-fold helpers?

thehogfather replied:
It means that line indentation is used as a suggestion for fold regions. Although this shouldn't take priority over range finders based on the current mode. So where there is conflict, the brace-fold-finder should take precedence. Having said that this might be another pref that could be better off disabled by default.

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -0,0 +1,352 @@
+        Initialise the extension
+    */
+    function init() {
+        if (CodeMirror.fold.combine || !prefs.getSetting("enabled")) {
```
When would `CodeMirror.fold.combine` be true?

thehogfather replied:
It would be truthy if the code-folding plugin is already loaded. This is mainly  for the extension in the registry- to prevent registering key bindings twice etc. I could not find an api to check that an extension has loaded. That code fragment can be removed without affecting the logic of initialisation.

peterflynn replied:
Ah, in that case -- core extensions are always loaded before user-installed extensions, so I think this logic is only needed in the registry copy, not here.

abose commented:
@thehogfather I was verifying the unit tests and some of the integration tests in the unit test setup is failing. 
Most is with this message- Cannot read property 'lastIndexOf' of undefined 
Could you take a look at debug>run tests>integration>all .

abose commented:
Code folding is enabled in all files by default, and in some files like txt files it looks a bit odd.
Would it be a good thing to enable code folding in only HTML,XML,js Etc. and disabled in all other file types?

thehogfather commented:
@abose wrt unit tests, I am unable to run them completely. They freeze with about 125 tests to go. Can you tell me the file throwing the error or show a stack trace? I remember spotting one a while back in one of the imported codeMirror range finders.

abose commented:
There is a memory issue with the unit tests currently, so we won't be able to run all tests in a single go.
So in order to run it, we have to the tests in individual sections.
First restart brackets; then open test suite, click on integration tab, click on all; it would start running integration tests. If brackets slows down, Restart brackets again to run tests.
In integration test, specifically the following test are failing[ FindInFiles(3 test cases),FindReplace - Integration(1),	InlineEditorProviders(3), QuickOpen(3)]. 
I am checking out these errors and the stack.

But Issue#10857 seems to be much more important than this. My understanding of the extension is limited; it might take a while for me to fix it. Could you take a look at the issue?

peterflynn commented:
@abose I took a quick look at those unit tests and it seems like the Health Report notification dialog might be causing problems in some of them (you can see it popping up in each test run, even though there's code that should suppress it).

The 3 QuickOpen failures might be different -- I can repro them on my old smart-autocompelte branch (which doesn't have code folding _or_ Health Report), so those failures might be caused by my recent PR there.  I can try to take a look tomorrow -- lmk if you find anything out before then.

abose commented:
The health notifications dialogs will be disabled when unit tests are nunning, but i guess that PR is still under review.
For the time being, it can be disabled by going to HealthDataNotification.js > AppInit.appReady()> 
params.parse();if (params.get("testEnvironment")) return;

redmunds commented:
There's a wiki page with [Tips and Tricks for Running Brackets Unit Tests](https://github.com/adobe/brackets/wiki/Running-Brackets-Unit-Tests) so just add any new tips there and share this link.

Pooria-H commented:
Why there is no option to disable this feature? It's ugly.

le717 commented:
@Pooria-H There is an option. In your `brackets.json` file, set `code-folding.enabled` to `false`. You can get more details on the [wiki](https://github.com/adobe/brackets/wiki/How-to-Use-Brackets#preferences).

ryanstewart commented:
Thanks @le717 !

Pooria-H commented:
@le717 Oh I didn't noticed. thanks
